cloud/pkg/synccontroller/objectsync.go

```diff
@@ -34,46 +36,59 @@ func (sctl *SyncController) manageObject(sync *v1alpha1.ObjectSync) {
 	ret, err := sctl.kubeclient.Resource(gvr).Namespace(sync.Namespace).Get(context.TODO(), sync.Spec.ObjectName, metav1.GetOptions{})
 	if err != nil || ret == nil {
 		klog.Errorf("failed to get obj(gvr:%v,namespace:%v,name:%v), %v", gvr, sync.Namespace, sync.Spec.ObjectName, err)
+		sendFailEvents(sync, err)
 		return
 	}
 
-	nodeName := getNodeName(sync.Name)
-
 	object, err = meta.Accessor(ret)
 	if err != nil {
+		klog.Errorf("failed to get obj with meta (gvr:%v,namespace:%v,name:%v), %v", gvr, sync.Namespace, sync.Spec.ObjectName, err)
+		sendFailEvents(sync, err)
 		return
 	}
-
 	syncObjUID := getObjectUID(sync.Name)
 	if syncObjUID != string(object.GetUID()) {
 		err = apierrors.NewNotFound(schema.GroupResource{
 			Group:    "",
 			Resource: sync.Spec.ObjectKind,
 		}, sync.Spec.ObjectName)
+		sendFailEvents(sync, err)
+		return
 	}
-
-	sendEvents(err, nodeName, sync, sync.Spec.ObjectKind, object.GetResourceVersion(), object)
+	sendEvents(sync, object.GetResourceVersion(), object)
 }
 
-func sendEvents(err error, nodeName string, sync *v1alpha1.ObjectSync, resourceType string,
-	objectResourceVersion string, obj interface{}) {
-	runtimeObj := obj.(runtime.Object)
-	if err := util.SetMetaType(runtimeObj); err != nil {
-		klog.Warningf("failed to set metatype :%v", err)
+func sendFailEvents(sync *v1alpha1.ObjectSync, err error) {
+	if err == nil {
+		return
 	}
-	if err != nil && apierrors.IsNotFound(err) {
+	nodeName := getNodeName(sync.Name)
+	if apierrors.IsNotFound(err) {
+		// build default obj
+		obj := &unstructured.Unstructured{}
+		obj.SetName(sync.Spec.ObjectName)
+		obj.SetUID(types.UID(getObjectUID(sync.Name)))
+		obj.SetNamespace(sync.Namespace)
+
 		//trigger the delete event
-		klog.Infof("%s: %s has been deleted in K8s, send the delete event to edge", resourceType, sync.Spec.ObjectName)
-		msg := buildEdgeControllerMessage(nodeName, sync.Namespace, resourceType, sync.Spec.ObjectName, model.DeleteOperation, obj)
+		klog.Infof("%s: %s has been deleted in K8s, send the delete event to edge", sync.Spec.ObjectKind, sync.Spec.ObjectName)
+		msg := buildEdgeControllerMessage(nodeName, sync.Namespace, sync.Spec.ObjectKind, sync.Spec.ObjectName, model.DeleteOperation, obj)
 		beehiveContext.Send(commonconst.DefaultContextSendModuleName, *msg)
-		return
 	}
+	// TODO Other error handling methods
+}
 
+func sendEvents(sync *v1alpha1.ObjectSync, objectResourceVersion string, obj interface{}) {
+	nodeName := getNodeName(sync.Name)
+	resourceType := sync.Spec.ObjectKind
```

Comment:
just question: shoud we do something about `sync.Spec.ObjectKind`?  
E.g `sync.Spec.ObjectKind` may be “Pod”， I think here need “pod” or “pods”.

Reply:
Yes, the `metamanager` will handle it. The database `meta` table stores kind.